include/api/wasmedge/wasmedge.hh

```diff
@@ -0,0 +1,707 @@
+    // In the AddProposal method -
+    // void AddProposal(const Proposal Prop)
+    // {
+    //   this.Conf.addProposal(Prop);
+    // }
+    // Configure Conf;
+  };
+
+  class WASMEDGE_CPP_API_EXPORT StatisticsContext {
+  public:
+    StatisticsContext();
+    ~StatisticsContext() = default;
+
+    uint64_t GetInstrCount();
+    double GetInstrPerSecond();
+    uint64_t GetTotalCost();
+
+    void SetCostTable(std::vector<uint64_t> CostArr, const uint32_t Len);
+    void SetCostLimit(const uint64_t Limit);
+    void Clear();
+
+  private:
+    // TODO
+  };
+
+  // >>>>>>>> WasmEdge Runtime >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
+
+  class WASMEDGE_CPP_API_EXPORT VM {
+  public:
+    VM(const ConfigureContext &ConfCxt, Store &StoreCxt);
+    ~VM() = default;
+
+    Result RegisterModuleFromFile(const String ModuleName,
+                                    const std::string &Path);
+    Result RegisterModuleFromBuffer(const String ModuleName,
+                  const std::vector<uint8_t> &Buf, const uint32_t BufLen);
+    Result RegisterModuleFromASTModule(const String ModuleName,
+                                      const ASTModule &ASTCxt);
+    Result RegisterModuleFromImport(const ModuleInstance &ImportCxt);
+
+    Result RunWasmFromFile(const std::string &Path, const String FuncName,
+                  const std::vector<Value> &Params, const uint32_t ParamLen,
+                  std::vector<Value> &Returns, const uint32_t ReturnLen);
+    Result RunWasmFromBuffer(const std::vector<uint8_t> &Buf,
+                  const uint32_t BufLen, const String FuncName,
+                  const std::vector<Value> &Params, const uint32_t ParamLen,
+                  std::vector<Value> &Returns, const uint32_t ReturnLen);
+    Result RunWasmFromASTModule(const ASTModule &ASTCxt, const String FuncName,
+                  const std::vector<Value> &Params, const uint32_t ParamLen,
+                  std::vector<Value> &Returns, const uint32_t ReturnLen);
+
+    std::unique_ptr<Async> AsyncRunWasmFromFile(const std::string &Path,
+                  const String FuncName, const std::vector<Value> &Params,
+                  const uint32_t ParamLen);
+    std::unique_ptr<Async> AsyncRunWasmFromBuffer(
+                  const std::vector<uint8_t> &Buf, const uint32_t BufLen,
+                  const String FuncName, const std::vector<Value> &Params,
+                  const uint32_t ParamLen);
+    std::unique_ptr<Async> AsyncRunWasmFromASTModule(const ASTModule &ASTCxt,
+                  const String FuncName, const std::vector<Value> &Params,
+                  const uint32_t ParamLen);
+
+    Result LoadWasmFromFile(const std::string &Path);
+    Result LoadWasmFromBuffer(const std::vector<uint8_t> &Buf,
+                  const uint32_t BufLen);
+    Result LoadWasmFromASTModule(const ASTModule &ASTCxt);
+
+    Result Validate();
+    Result Instantiate();
+
+    Result Execute(const String FuncName, const std::vector<Value> &Params,
```

Comment:
The `ParamLen` for C API is because the C API doesn't have `vector`.
With `std::vector`, the length argument is not needed.